### PR TITLE
Jetpack CLI: Add ability to generate plugin using a function starter-plugin as a template

### LINKED ISSUE #23441: Improve plugin scaffolder
A few tasks oriented to streamlining the process of creating a fully functional plugin that's able to connect the site and show My Jetpack as soon as it's been generated.

jetpackcli

Most of these tasks are being handled by #23434. Props @leogermani 🙌 

* [x] Make it load the PHP autoloader 
* [ ] Include basic internal monorepo dependencies
  * [x] Include My Jetpack as dependency
  * [x] Include Jetpack autoloader as dependency
  * [x] Include Jetpack\Admin_UI as dependency
  * [x] Include Jetpack\Assets as dependency
  * [x] Include composer plugin to dependencies
* [x] Include PHPCS config file
* [x] Have the main plugin file file be `jetpack-xxxx.php` instead of `xxxx.php` 
* [x] Have the first class file be `class-jetpack-xxxx.php` instead of `example.php` 
* [x] Include  WordBless `wordpress` dir in .gitignore
* [x] Include  `build` in .gitignore
* [x] Include  `.cache` in .gitignore* 
* [x] Include Admin Page structure
  * [x] Include `src/js/index.js`.
  * [x] Include `components/admin-page`.
  * [x] Include JS tooling for Admin Page
    * [x] Provide `.eslintrc.js`
    * [x] Provide `babel.config.js`
    * [x] Provide `webpack.config.js`
    * [x] Provide `package.json`
* [x] Provide `build-development`, `build-production` and `watch` scripts in composer.json
* [x] Include `Jetpack\Config` as dependency.
  * [x] Include `Jetpack\Identity_Crisis` and `Jetpack\Sync as dependency` as they're depedency of Config.
* [x] Added ability to connect and connected/disconnected screens on admin page.
* [x] Include proper `.gitattributes` to have  vendor and jetpack_vendor dirs were present on the action builds.
* [x] Include autoloader-suffix to composer.json

### PR DESCRIPTION
**Pull request by leogermani**
Closes #23441
#### Changes proposed in this Pull Request:

This PR creates the Jetpack Starter Plugin and adds a new option to the generate plugin CLI command to build a new plugin using it as a template.

#### Jetpack product discussion

N/A

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Run `jetpack generate` CLI command
* Say you are generating a plugin
* When asked, choose the option to use the Starter plugin as a template
* When done, build the newly generated plugin with `jetpack build plugins/your-new-plugin`
* Test the plugin
* Verify that all names, slugs, class names and constants were renamed to match your new plugin name
* Test it with different plugin names, containing `-`, `_` and `.` in their names

Known issues
. [SOLVED] There's a hardcoded `autoloader suffix` in composer.json
. [SOLVED} When trying to commit, I got a babel error saying I needed to add a resolver for `jsx`, even though it is building successfully
. in `searchReplaceInFolder` there's a recursive call to itself that is not async. Could it generate problems?
. what else?

### CODE REVIEW

**Comment by kraftbj** [comment]
What's the advantage of having the starter plugin like this vs putting it into the plugin skeleton within the CLI, or as a second skeleton within the CLI?

**Comment by retrofox** [comment]
> What's the advantage of having the starter plugin like this vs putting it into the plugin skeleton within the CLI, or as a second skeleton within the CLI?

Sorry for intruding here, but could it one advantage to be simpler to improve the plugin code, features, etc from the angle of developers not specialized on the CLI tool?

**Comment by leogermani** [comment]
> What's the advantage of having the starter plugin like this vs putting it into the plugin skeleton within the CLI, or as a second skeleton within the CLI?

It's easier to iterate on the starter plugin as we can actually build it and test it live.

**Comment by oskosk** [comment]
Pushed 6a6a5e86dec7f2b7af050037edf0f0884888ac13 to add `.eslintrc.js` file and address the lack of JSX parsing in index.js when running the lint commands (Some check was failing). Confirmed the text replacer worked fine on it

File: projects/plugins/starter-plugin/src/class-jetpack-starter-plugin.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+	}
+
+	/**
+	 * Render the initial state into a JavaScript variable.
+	 *
+	 * @return string
+	 */
+	public function render_initial_state() {
+		return 'var jetpackSocialInitialState=JSON.parse(decodeURIComponent("' . rawurlencode( wp_json_encode( $this->initial_state() ) ) . '"));';
```
Comment (anomiex) [inline projects/plugins/starter-plugin/src/class-jetpack-starter-plugin.php:100]: This needs updating too.

**Review by anomiex — COMMENTED** [review]
Overall this seems more of a small plugin people might want to look at as an example than a usable skeleton.

**Comment by kraftbj** [comment]
Okay. I'd think once the iteration is no longer rapidly happening, we'd move it to the CLI skeleton if that's the intent of the code (as opposed as an example to @anomiex point).

The goal of the CLI generation is not only copy/paste from the skeleton but update values to match the new intended plugin. Also having the skeletons and the sample plugin as separately located skeletons will mean multiple places to update when needed and extra CI runs as it's seen as a regular plugin.

File: projects/plugins/starter-plugin/composer.json [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			"options": {
+				"monorepo": true
+			}
+		}
+	],
+	"minimum-stability": "dev",
+	"prefer-stable": true,
+	"extra": {
+		"mirror-repo": "Automattic/jetpack-starter-plugin-plugin",
+		"release-branch-prefix": "social",
```
[the remark this answers is not in the document]
Comment (leogermani) [reply projects/plugins/starter-plugin/composer.json]: addressed

File: projects/plugins/starter-plugin/composer.json [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+		"wp-plugin-slug": "jetpack-starter-plugin"
+	},
+	"config": {
+		"allow-plugins": {
+			"roots/wordpress-core-installer": true,
+			"automattic/jetpack-autoloader": true,
+			"automattic/jetpack-composer-plugin": true
+		},
+		"autoloader-suffix": "c4802e05bbcf59fd3b6350e8d3e5482c_socialⓥ0_1_0_alpha"
```
[the remark this answers is not in the document]
Comment (leogermani) [reply projects/plugins/starter-plugin/composer.json]: addressed

File: projects/plugins/starter-plugin/.eslintrc.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
+	extends: [ '../../../tools/js-tools/eslintrc/react.js' ],
+	parserOptions: {
+		babelOptions: {
+			configFile: require.resolve( './babel.config.js' ),
+		},
+	},
+	rules: {
+		// Enforce the use of the jetpack-social textdomain.
```
[the remark this answers is not in the document]
Comment (leogermani) [reply projects/plugins/starter-plugin/.eslintrc.js]: addressed

File: projects/plugins/starter-plugin/package.json [hunk 5]
```diff
@@ -0,0 +1,55 @@
+{
+	"private": true,
+	"description": "Share your site’s posts on several social media networks automatically when you publish a new post.",
```
[the remark this answers is not in the document]
Comment (leogermani) [reply projects/plugins/starter-plugin/package.json]: addressed

File: projects/plugins/starter-plugin/package.json [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	"repository": {
+		"type": "git",
+		"url": "https://github.com/Automattic/jetpack.git"
+	},
+	"license": "GPL-2.0-or-later",
+	"author": "Automattic",
+	"scripts": {
+		"build": "pnpm run clean && pnpm run build-client",
+		"build-client": "pnpx webpack",
+		"build-concurrently": "pnpm run clean && pnpx concurrently 'pnpm:build-client' 'pnpm:build-php'",
```
[the remark this answers is not in the document]
Comment (leogermani) [reply projects/plugins/starter-plugin/package.json]: addressed

File: tools/cli/commands/generate.js [hunk 7]
```diff
@@ -257,6 +284,75 @@ export async function generateProject(
...
+
+	// Delete untracked folders.
+	const untrackedFolders = [
+		'/vendor',
+		'/jetpack_vendor',
+		'/node_modules',
+		'/build',
+		'/.cache',
+		'/wordpress',
+	];
```
[the remark this answers is not in the document]
Comment (leogermani) [reply tools/cli/commands/generate.js]: addressed

**Comment by anomiex** [comment]
> One thing I noticed is that the symlink inside the docker container for the new plugin to show up in the UI is not being created... Was it being created before? @anomiex ?

The symlink creation happens during the container startup. Have you stopped and restarted your container while having this checked out?

**Comment by leogermani** [comment]
> The symlink creation happens during the container startup. Have you stopped and restarted your container while having this checked out?

It works. But it means I have to restart the containers after running the generate command. Should we add a note at the script output maybe?

**Review by oskosk — APPROVED** [review]
LGTM! Excellent work!

**Comment by oskosk** [comment]
r242395-wpcom

**Pull request merged**